internal/planner/planner.go

```diff
@@ -1492,6 +1499,37 @@ func (p *Planner) planRefData(virtual *ruletrie, base *baseptr, ref ast.Ref, ind
 		return p.planRefDataExtent(virtual, base, iter)
 	}
 
+	// On the first iteration, we check if this can be optimized using a
+	// CallDynamicStatement
+	// NOTE(sr): we do it on the first index because later on, the recursion
+	// on subtrees of virtual already lost parts of the path we've taken.
+	if index == 1 {
+		rulesets, stmts, locals, index, optimize := p.optimizeLookup(virtual, ref)
+		if optimize {
```

Comment:
I could imagine it being useful to have the planner emit debug messages like compile/compile.go does for some of the optimizations it performs. Otherwise it becomes tricky to understand why the optimization hasn't kicked in.